Bluetooth: Controller: Use defines for TRX and IFS Radio End Events

cvinayak opened the pull request:
Use defines for Transmit/Receive and Inter-frame spacing
related Radio End events. This will ease some spaghetti of
conditional compilations for single timer and direction
finding support.

cvinayak committed:
Bluetooth: Controller: Fixup align DDPI usage with nRF54L power domains

Fixup commit 031a17c2716e ("Bluetooth: Controller: Align
DDPI usage with nRF54L power domains").

Align DDPI usage with nRF54L series power domains and
available DDPI configurations.

Use indices below 4 when needing to bridge with GPIO,
use indices below 8 when needing to bridge with MCU domain
like CCM and AAR; and use indices above 7 to bridge with
Peripheral domain.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Redundant TIMER_CLEAR_PPI disable

HAL_SW_SWITCH_TIMER_CLEAR_PPI is same as
HAL_SW_SWITCH_GROUP_TASK_ENABLE_PPI, hence the single PPI
suffice being disabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
Code diff tested on nrf5340dk:
```
diff --git a/samples/bluetooth/direction_finding_connectionless_rx/src/main.c b/samples/bluetooth/direction_finding_connectionless_rx/src/main.c
index 3c5304e2c0d..5e7a7914c99 100644
--- a/samples/bluetooth/direction_finding_connectionless_rx/src/main.c
+++ b/samples/bluetooth/direction_finding_connectionless_rx/src/main.c
@@ -302,7 +302,7 @@ static void enable_cte_rx(void)
        int err;
 
        const struct bt_df_per_adv_sync_cte_rx_param cte_rx_params = {
-               .max_cte_count = 5,
+               .max_cte_count = 3,
 #if defined(CONFIG_BT_DF_CTE_RX_AOA)
                .cte_types = BT_DF_CTE_TYPE_ALL,
                .slot_durations = 0x2,
```

Log output for (not single timer use):
```
[2025-05-16 12:05:46.649] [DEVICE]: FF:6C:F3:56:DB:E0 (random), AD evt type 5, Tx Pwr: 127, RSSI -76 DF Connectionless Beacon App C:0 S:0 D:0 SR:0 E:1 Prim: LE 1M, Secn: LE 2M, Interval: 0x03c0 (1200 ms), SID: 0
[2025-05-16 12:05:46.665] [DEVICE]: 72:5A:5B:2D:2B:5D (random), AD evt type 0, Tx Pwr: 127, RSSI -99  C:1 S:1 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.679] [DEVICE]: 0A:A4:6A:B9:10:4F (random), AD evt type 3, Tx Pwr: 127, RSSI -94  C:0 S:0 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.693] success. Found periodic advertising.
[2025-05-16 12:05:46.697] Creating Periodic Advertising Sync...failed (err -16)
[2025-05-16 12:05:46.701] Waiting for periodic sync...
[2025-05-16 12:05:46.759] [DEVICE]: 2C:BE:EB:E6:14:CE (public), AD evt type 0, Tx Pwr: 127, RSSI -98  C:1 S:1 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.773] [DEVICE]: 7D:59:DF:97:07:EC (random), AD evt type 0, Tx Pwr: 127, RSSI -84  C:1 S:1 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.787] [DEVICE]: 44:38:F3:CE:FB:A9 (random), AD evt type 0, Tx Pwr: 127, RSSI -95  C:1 S:1 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.801] [DEVICE]: 44:38:F3:CE:FB:A9 (random), AD evt type 4, Tx Pwr: 127, RSSI -94  C:1 S:1 D:0 SR:1 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.817] [DEVICE]: 26:A0:33:D0:4D:77 (random), AD evt type 3, Tx Pwr: 127, RSSI -88  C:0 S:0 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.839] [DEVICE]: 1B:FC:32:4B:5C:C6 (random), AD evt type 3, Tx Pwr: 127, RSSI -97  C:0 S:0 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.877] [DEVICE]: 15:79:BD:51:4E:66 (random), AD evt type 3, Tx Pwr: 127, RSSI -67  C:0 S:0 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.891] [DEVICE]: 78:DE:F8:F4:3E:5C (random), AD evt type 0, Tx Pwr: 127, RSSI -97  C:1 S:1 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.905] [DEVICE]: 3F:36:18:E6:93:5E (random), AD evt type 3, Tx Pwr: 127, RSSI -82  C:0 S:0 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.919] [DEVICE]: 06:CE:F7:62:7C:7A (random), AD evt type 3, Tx Pwr: 127, RSSI -89  C:0 S:0 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.933] PER_ADV_SYNC[0]: [DEVICE]: FF:6C:F3:56:DB:E0 (random) synced, Interval 0x03c0 (1200 ms), PHY LE 2M
[2025-05-16 12:05:46.942] [DEVICE]: A3:CD:F2:A1:31:5E (public), AD evt type 0, Tx Pwr: 127, RSSI -89 IDM-A1315E C:1 S:1 D:0 SR:0 E:0 Prim: LE 1M, Secn: No packets, Interval: 0x0000 (0 ms), SID: 255
[2025-05-16 12:05:46.957] success. Periodic sync established.
[2025-05-16 12:05:46.960] Enable receiving of CTE...
[2025-05-16 12:05:46.963] success. CTE receive enabled.
[2025-05-16 12:05:46.966] Scan disable...Success.
[2025-05-16 12:05:46.968] Waiting for periodic sync lost...
[2025-05-16 12:05:48.123] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -660
[2025-05-16 12:05:48.134] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -730
[2025-05-16 12:05:48.145] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -700
[2025-05-16 12:05:49.323] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -680
[2025-05-16 12:05:49.334] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -680
[2025-05-16 12:05:49.345] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -700
[2025-05-16 12:05:50.523] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -650
[2025-05-16 12:05:50.534] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -750
[2025-05-16 12:05:50.545] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -680
[2025-05-16 12:05:51.723] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -840
[2025-05-16 12:05:51.734] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -680
[2025-05-16 12:05:51.745] CTE[0]: samples type: 8 bits int, samples count 45, cte type AOD 2 [us], slot durations: 2 [us], packet status CRC OK, RSSI -700

```


single timer and direction finding does not seem to have been ever supported to my knowledge, hence this PR does not break anything for nRF53 (definitely not for nrf52 that uses PPI, as this PR only concerns DPPI).

I have tested samples/bluetooth/periodic_sync  with and without single timer use, and that has no regression in nrf53.

Hence, functionally this PR is good. (single timer and DF support has to be in a separate PR).

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi_resources.h:
```diff
@@ -109,11 +109,11 @@
  * bit or CRC is marked by END event.
  */
-#if defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER)
+#if defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER) && !defined(CONFIG_BT_CTLR_DF)
```
I wonder if `defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER) && !defined(CONFIG_BT_CTLR_DF)` or `!defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER) && defined(CONFIG_BT_CTLR_DF)` could be put into a new promptless Kconfig option or otherwise stored as a `#define` or something with a meaningful name?

cvinayak replied:
Sure, will give a try.

cvinayak replied:
Each conditional compilation logic is a different combination (catering to nrf52-ppi or nrf53/54-DPPI or single timer use), using a Kconfig or #define needs significant refactoring of the implementation. As the implementation has incremental been updated over the years to support changes in SoC h/w, it is now a spaghetti.

Thalley replied:
Yeah, it's pretty hard to follow that many conditionals (some of which are even in other conditionals). 

I can approve for now, but I think some effort needs to be done for this, as it significantly reduces maintainability.

cvinayak committed:
Bluetooth: Controller: Use defines for TRX and IFS Radio End Events

Use defines for Transmit/Receive and Inter-frame spacing
related Radio End events. This will ease some spaghetti of
conditional compilations for single timer and direction
finding support.

Direction Finding is not yet supported when using single
timer software tIFS switching.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Do not clear timer clear subscription until done

Do not clear timer clear subscription until the radio event
is done. This is required for single timer switching when
periodic chains are to be received by LLL scheduling. Each
PDU in the train requires the timer to be cleared when the
parent PDU has been received such that reception of
auxiliary PDU's radio_tmr_start_us() can use the relative
offset since last PDU reception.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix incorrect aux offset using single timer

Fix incorrect aux offset value in the aux pointer in the
primary advertising channel PDUs when using single timer
tIFS switching feature in nRF54Lx.

End time was not captured when only Tx follow by Tx happens
in the primary advertising channels. End time needs to be
always captured when using single timer so that elapsed
packet timer value is correctly accumulated, for use to
setup subsequent radio transmission or reception inside the
same radio event.

Relates to commit 373dc0db71d1 ("Bluetooth: Controller:
Single timer use Extended Advertising nRF54L support").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Enable testing periodic adv chains on nRF54lx

Enable testing periodic advertising chains on nrf54l15bsim
target to cover the Controller's use of single timer for
packet and tIFS switching.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
